### ISSUE TITLE
Add warnings to, or deprecate, `MappingQ`

### ISSUE THREAD

**Issue by drwells** [dealii/dealii#4462]
As discussed in #3874: this class has a number of caveats. At a minimum I think that we should describe the one case in which one can actually achieve something useful with this class (quadratic elements, set all interior face manifold ids to straight) and put an explanatory warning describing what happens when one tries to do what I did in #3874 (i.e., we are limited to third order convergence).

Given these severe limitations I think that it would be better to deprecate the class. The `deal2lkit` codes only ever use the quadratic or linear versions of this (which are fine) and `ASPECT` always uses the higher order mapping on interior cells (also OK): there are a few other user codes that use this and may not realize that it does not do what they want it do do. @koecher It looks like nearly every project of yours uses this class. Have you noticed these types of convergence problems?

**Comment by koecher** [comment]
@drwells, yes I'm using MappingQ in a lot of my codes, but I do not use meshes with curved boundaries or meshes with curved interior faces that much.
In the case you described, meshes with quadratic elements all interior faces are "straight", it works without problems together with higher order methods.
The reason for the inclusion draws back to my first year with deal.II, I think in most of the cases it is not necessary.

**Comment by drwells** [comment]
We should close this issue and keep MappingQ. In particular (for the sake of record): the culprit behind the reduced rates of convergence was the Laplace-based algorithm for placing new points, not the choice of mapping.

**Comment by kronbichler** [comment]
There is a catch with this class which I wanted to raise: One silently gets incompatible meshes if there are volume manifolds that deform cells and uses `MappingQ`. For example when using `MappingQ` with a `hyper_shell` mesh and its associated boundary. We should try to throw an error in that case, don't you think?

**Comment by kronbichler** [comment]
To illustrate my point, consider this simple code:
```
    const unsigned int dim=2;

    Triangulation<dim> tria;
    GridGenerator::hyper_shell(tria, Point<dim>(), 0.8, 1., 3);
    tria.refine_global(2);


    FE_Q<dim> fe(1);
    DoFHandler<dim> dofh(tria);
    dofh.distribute_dofs(fe);
    Vector<double> vec(dofh.n_dofs());

    DataOut<dim> data_out;
    data_out.attach_dof_handler(dofh);
    data_out.add_data_vector(vec, "vec");
    MappingQ<dim> mapping(3);
    data_out.build_patches(mapping, 2, DataOut<dim>::curved_inner_cells);

    std::ofstream file("grid.vtk");
    data_out.write_vtk(file);
```
If I, as a naive user, run this, I will get the following weird mesh (for computing things):
![wrong](https://user-images.githubusercontent.com/8276200/40617575-ff5736a6-628e-11e8-8d81-188058772ece.png)
whereas everyone would expect this one (which I get by simply replacing `MappingQ` by `MappingQGeneric`:
![correct](https://user-images.githubusercontent.com/8276200/40617588-0fb81704-628f-11e8-94fe-63a8bb1aaa57.png)

Now of course we developers know how to set things up properly, but one needs to read the documentation carefully to use the right mesh. What is more, it is usually not completely trivial to see this defect - one rather gets weird results that are not completely wrong - i.e., the worst situation to debug I can imagine.

**Comment by kronbichler** [comment]
What I wanted to say: Now that we have automatic manifold attachment, we should either raise an exception or, even better, silently go back to what `MappingQGeneric` does as soon as we detect a manifold that is not the flat manifold.

**Comment by drwells** [comment]
The issue with patches is a good example. I want to try a few things to convince myself that we should raise an error.

**Comment by starki0815** [comment]
The topic is a bit old ... but the problem seems to be still there. I just ran into exactly the problem illustrated by the example of @kronbichler and it took me almost a day to figure out what's wrong (below is an example showing that the solution may not converge at all because of the gaps/overlaps in the geometry). So, I definitely agree that there should be an exception raised if anything else than a flat manifold is attached to the interior.

```cpp
#include <deal.II/grid/tria.h>
#include <deal.II/dofs/dof_handler.h>
#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/tria_accessor.h>
#include <deal.II/grid/tria_iterator.h>
#include <deal.II/dofs/dof_accessor.h>
#include <deal.II/fe/fe_q.h>
#include <deal.II/fe/fe_system.h>
#include <deal.II/dofs/dof_tools.h>
#include <deal.II/fe/fe_values.h>
#include <deal.II/base/quadrature_lib.h>
#include <deal.II/base/function.h>
#include <deal.II/numerics/vector_tools.h>
#include <deal.II/numerics/matrix_tools.h>
#include <deal.II/lac/vector.h>
#include <deal.II/lac/full_matrix.h>
#include <deal.II/lac/sparse_matrix.h>
#include <deal.II/lac/dynamic_sparsity_pattern.h>
#include <deal.II/numerics/data_out.h>
#include <deal.II/grid/manifold_lib.h>
#include <deal.II/fe/mapping_q.h>
#include <deal.II/lac/sparse_direct.h>
#include <fstream>
#include <iostream>
#include <math.h>

using namespace dealii;
using namespace std;

// exact solution of a Laplace problem on a two-dimensional domain bounded by two concentric circles
// (radii r=0.5 and r=1.0, respectively) which are centered at the origin
//
// value of the potential prescribed to zero at the inner boundary and to 1.0 at the outer boundary
//
// components 0 and 1 are the negative of the gradient of the potential satisfying the Laplace equation
// component 2 is the potential satisfying the Laplace equation
class ExactSolution : public Function<2>
{
	public:
		ExactSolution() : Function<2>(3) {}
	double value (const Point<2>&p, const unsigned int component) const
	{
		if(component==0)
			return -1./log(2.) * p(0) / (p(0)*p(0)+p(1)*p(1));
		else if(component==1)
			return -1./log(2.) * p(1) / (p(0)*p(0)+p(1)*p(1));
		else
			return 1./log(2.) * log( sqrt(p(0)*p(0)+p(1)*p(1)) / 0.5 );
	}
};

//Mixed Laplace class (essentially copied and pasted from the deal.ii tutorials)
class MixedLaplace
{
	public:
		MixedLaplace(const Mapping<2>& mapping);
		void run(const unsigned int refinements);
	private:
		void make_grid(const unsigned int refinements);
		void setup_system();
		void assemble_system();
		void solve();
		void output_results(const unsigned int file_index) const;
		void difference_to_exact_solution() const;
		Triangulation<2>			triangulation;
		FESystem<2>					fe;
		DoFHandler<2>				dof_handler;
		SparsityPattern				sparsity_pattern;
		SparseMatrix<double>		system_matrix;
		Vector<double>				solution;
		Vector<double>				system_rhs;
		const SphericalManifold<2>	manifold;
		const Mapping<2>&			mapping;
};

MixedLaplace::MixedLaplace(const Mapping<2>& mapping)
	:
	//use quadratic elements for gradients of potential and linear elements for potential itself in order to satisfy inf-sup
	fe(FE_Q<2>(2), 2, FE_Q<2>(1), 1),
	dof_handler(triangulation),
	mapping(mapping)
{}

void MixedLaplace::make_grid(const unsigned int refinements)
{
	GridGenerator::hyper_shell(triangulation, Point<2>(), 0.5, 1, 8);
	for(const auto& cell : triangulation.active_cell_iterators())
	{
		for(unsigned int face=0; face<GeometryInfo<2>::faces_per_cell; ++face)
		{
			if(cell->face(face)->at_boundary()){
				const Point<2> point=cell->face(face)->center();
				double r=point.distance(Point<2>());
				if(r<0.75)
					cell->face(face)->set_boundary_id(0);
				else
					cell->face(face)->set_boundary_id(1);
			}
		}

	}
	triangulation.set_all_manifold_ids(0);
	triangulation.set_manifold(0, manifold);
	triangulation.refine_global(refinements);
}

void MixedLaplace::setup_system()
{
	dof_handler.distribute_dofs(fe);
	DynamicSparsityPattern dsp(dof_handler.n_dofs());
	DoFTools::make_sparsity_pattern(dof_handler, dsp);
	sparsity_pattern.copy_from(dsp);
	system_matrix.reinit (sparsity_pattern);
	solution.reinit(dof_handler.n_dofs());
	system_rhs.reinit(dof_handler.n_dofs());
}
void MixedLaplace::assemble_system()
{
	QGauss<2> quadrature_formula(3);
    QGauss<1> face_quadrature_formula(3);
    FEValues<2> fe_values(	mapping, fe, quadrature_formula,
    						update_values    | update_gradients |
							update_quadrature_points  | update_JxW_values);
    FEFaceValues<2> fe_face_values(	mapping, fe, face_quadrature_formula,
    								update_values    | update_normal_vectors |
									update_quadrature_points  | update_JxW_values);
	const unsigned int	dofs_per_cell = fe.dofs_per_cell;
	const unsigned int	n_q_points = quadrature_formula.size();
	const unsigned int n_face_q_points = face_quadrature_formula.size();
	FullMatrix<double>	cell_matrix(dofs_per_cell, dofs_per_cell);
	Vector<double>		cell_rhs(dofs_per_cell);
	vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
	const FEValuesExtractors::Vector velocities(0);
	const FEValuesExtractors::Scalar pressure(2);

	for(const auto &cell: dof_handler.active_cell_iterators())
	{
		cell_matrix = 0;
		cell_rhs = 0;
		fe_values.reinit(cell);
		//assemble volume terms
		for (unsigned int q=0; q<n_q_points; ++q)
			for (unsigned int i=0; i<dofs_per_cell; ++i)
			{
				const Tensor<1,2> 	phi_i_u		= fe_values[velocities].value(i, q);
				const double		div_phi_i_u = fe_values[velocities].divergence(i, q);
				const double		phi_i_p		= fe_values[pressure].value(i, q);
				for (unsigned int j=0; j<dofs_per_cell; ++j)
				{
					const Tensor<1,2>	phi_j_u	= fe_values[velocities].value(j, q);
					const double		div_phi_j_u = fe_values[velocities].divergence(j, q);
					const double		phi_j_p     = fe_values[pressure].value(j, q);
					cell_matrix(i,j) += 	(phi_i_u * phi_j_u
											- div_phi_i_u * phi_j_p
											- phi_i_p * div_phi_j_u)
											* fe_values.JxW(q);
				}
			}
		//assemble boundary term to implement that potential is 1 at outer boundary
		for(unsigned int face_n=0; face_n<GeometryInfo<2>::faces_per_cell; ++face_n)
			if (cell->at_boundary(face_n) && cell->face(face_n)->boundary_id()==1)
			{
				fe_face_values.reinit (cell, face_n);
				for (unsigned int q=0; q<n_face_q_points; ++q)
					for (unsigned int i=0; i<dofs_per_cell; ++i)
						cell_rhs(i) +=	-(fe_face_values[velocities].value(i, q) *
										fe_face_values.normal_vector(q) *
										fe_face_values.JxW(q));
			}
		//distribute to global matrix and rhs
		cell->get_dof_indices(local_dof_indices);
		for(unsigned int i=0; i<dofs_per_cell; ++i)
			for (unsigned int j=0; j<dofs_per_cell; ++j)
				system_matrix.add(	local_dof_indices[i],
									local_dof_indices[j],
									cell_matrix(i,j));
			for (unsigned int i=0; i<dofs_per_cell; ++i)
				system_rhs(local_dof_indices[i]) += cell_rhs(i);
		}
}

void MixedLaplace::solve()
{
	SparseDirectUMFPACK  A_direct;
	A_direct.initialize(system_matrix);
	A_direct.vmult(solution, system_rhs);
}

void MixedLaplace::output_results(const unsigned int file_index) const
{
	vector<string> solution_names(2, "u");
	solution_names.emplace_back("p");
	vector<DataComponentInterpretation::DataComponentInterpretation> interpretation(2, DataComponentInterpretation::component_is_part_of_vector);
	interpretation.push_back(DataComponentInterpretation::component_is_scalar);
	DataOut<2> data_out;
	data_out.add_data_vector (dof_handler, solution, solution_names, interpretation);
	data_out.build_patches(2);
	ofstream output("solution_" + 	Utilities::int_to_string(file_index) + ".vtk");
	data_out.write_vtk(output);
	cout << "  Solution written to=" << "solution_" + 	Utilities::int_to_string(file_index) + ".vtk" << endl;
}

void MixedLaplace::difference_to_exact_solution() const
{
	Vector<double> cell_differences(triangulation.n_active_cells());
	VectorTools::integrate_difference(mapping, dof_handler, solution, ExactSolution(), cell_differences, QGauss<2>(3), VectorTools::NormType::Linfty_norm);
	double e=VectorTools::compute_global_error 	(triangulation, cell_differences, VectorTools::NormType::Linfty_norm);
	cout << "  Error in L_infty norm=" << e << endl;
}

void MixedLaplace::run(const unsigned int refinements)
{
	make_grid(refinements);
	setup_system();
	assemble_system();
	solve();
	output_results(refinements);
	difference_to_exact_solution();
}

int main ()
{
	//if that's changed to MappingQGeneric, everything works fine
	const MappingQ<2> mapping(2);

	//number of refinement steps (problem is solved for each refinement step in order to analyze the convergence properties)
	const unsigned int n_refinements=4;

	//do the computations
	for(unsigned int i1=0; i1<n_refinements+1; ++i1){
		MixedLaplace laplace_problem(mapping);
		cout << "Number of refinements: " << i1 << ":" << endl;
		laplace_problem.run(i1);
	}
	return 0;
}
```

**Comment by drwells** [comment]
This might actually be a separate issue: I tried this code and set, after grid refinement, all non-boundary manifold ids to `numbers:;flat_manifold_id` and still only got first order convergence with `MappingQ`. This doesn't seem right.

**Comment by starki0815** [comment]
That's interesting. The same happens with MappingQGeneric (ok, that's to be expected in this case). Still, I think that the issue discussed above is also present in my example. But your observation indicates that there is some additional issue. In fact, the convergence rates are better if numbers::flat_manifold_id is assigned everywhere after refinement (including boundary faces)
I don't have an idea what causes this behavior.

**Comment by kronbichler** [comment]
@drwells how exactly did you set up that test? When I run with flat manifold ids in the interior (first setting all manifold ids to the flat one, then setting the ones on the boundary to `0`), then I get the same convergence as before, albeit with higher absolute error values. I get `0.0353491, 0.0142041, 0.00554808` as the last three error values, compared to `MappingQGeneric` with all-spherical manifold that delivers `0.0177647, 0.00636167, 0.00193966` and the erroneous case with `MappingQ` that gives me `0.164709, 0.0895128, 0.046225`. So to me this indeed looks like the original problem.

As a side node @starki0815 in case you don't know yet, to get good-looking output you probably want to set `data_out.build_patches(mapping, 2, DataOut<2>::curved_inner_cells);`.

**Comment by starki0815** [comment]
@kronbichler I think, you have to do more refinements to see the problem.

For the case with flat manifold in the interior and spherical manifold only at all boundary faces, I get
0.5 0.038747
0.25 0.0704014
0.125 0.0364466
0.0625 0.0144047
0.03125 0.00557172
0.015625 0.00247581
0.0078125 0.00124789
0.00390625 0.000626497
as output, which clearly indicates first order convergence. Actually, these results are slightly different than yours @kronbichler and I am wondering what causes the difference. After refinement, I do the following:
```cpp
	for(const auto& cell : triangulation.active_cell_iterators())
	{
		cell->set_all_manifold_ids(numbers::flat_manifold_id);
		for(unsigned int f=0; f<4; ++f){
			if(cell->face(f)->at_boundary()){
				cell->face(f)->set_manifold_id(0);
			}
		}
	}
```

For the case with spherical manifold everywhere, I get
0.5 0.038747
0.25 0.0442336
0.125 0.0177647
0.0625 0.00636167
0.03125 0.00193966
0.015625 0.000538904
0.0078125 0.000142291
0.00390625 3.65764e-05
as output, which seems to be second order convergence.

Thanks for the hint ... I had already learned about it from your above example ;)

**Comment by starki0815** [comment]
What I am wondering meanwhile is whether we can actually expect second order convergence for the case with the transition between spherical manifold on the boundary and flat manifold in the interior. Is there any reference, where it is proven that the transition between two different manifold descriptions within a single layer of elements by means of transfinite interpolation doesn't affect the convergence rates? I don't know much about error estimation, but intuitively I wouldn't be surprised if the fact that we have a transition in the manifold description within a single layer of elements would affect convergence.

I have also attached two pictures showing a close-up of the magnitude of the vector-valued variable for the case with a transition between manifolds and for the case with spherical manifold everywhere, respectively. No matter, how fine the mesh is, there is always an oscillating behavior of the solution for the former case, whereas there is no such oscillation for the latter (and I suspect that this is related to the fact that there is a transition in the manifold description within a single layer of elements).

Another thing which still worries me is the fact that @kronbichler got slightly different numbers than me for the case with manifold transition. So, @kronbichler can you post the code you have used and the version of the library?

![solu_flat_interior](https://user-images.githubusercontent.com/45016248/55185964-2b9e7b80-519e-11e9-8c80-8472deb7aeae.png)
![solu_curved_interior](https://user-images.githubusercontent.com/45016248/55185975-2f320280-519e-11e9-8b34-28ab3630dd7c.png)

**Comment by kronbichler** [comment]
I will look it up soon. Regarding the necessary ingredients for convergence, on standard problems (say Laplacian), you can read some of the discussions we had for the PR #5386.

**Comment by kronbichler** [comment]
In the end, my numbers also tend to the first order convergence rate. I guess this depends on the particular equation as the typical primal formulations should work well. Even though not as good as a transfinite interpolation manifold, for example, they still gave good convergence at least in my tests, see also #5386.
Here is the output with two more cycles: `0.038747 0.0704014 0.0353491 0.0142041 0.00554808 0.00262958 0.00132288 0.000665619` (but basically the same if I do not reset the inner faces to flat manifolds).

And here is my code, now with the `MappingQGeneric` setup that also only gives first order convergence:
```
#include <deal.II/grid/tria.h>
#include <deal.II/dofs/dof_handler.h>
#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/tria_accessor.h>
#include <deal.II/grid/tria_iterator.h>
#include <deal.II/dofs/dof_accessor.h>
#include <deal.II/fe/fe_q.h>
#include <deal.II/fe/fe_system.h>
#include <deal.II/dofs/dof_tools.h>
#include <deal.II/fe/fe_values.h>
#include <deal.II/base/quadrature_lib.h>
#include <deal.II/base/function.h>
#include <deal.II/numerics/vector_tools.h>
#include <deal.II/numerics/matrix_tools.h>
#include <deal.II/lac/vector.h>
#include <deal.II/lac/full_matrix.h>
#include <deal.II/lac/sparse_matrix.h>
#include <deal.II/lac/dynamic_sparsity_pattern.h>
#include <deal.II/numerics/data_out.h>
#include <deal.II/grid/manifold_lib.h>
#include <deal.II/fe/mapping_q.h>
#include <deal.II/lac/sparse_direct.h>
#include <fstream>
#include <iostream>
#include <math.h>

using namespace dealii;
using namespace std;

// exact solution of a Laplace problem on a two-dimensional domain bounded by two concentric circles
// (radii r=0.5 and r=1.0, respectively) which are centered at the origin
//
// value of the potential prescribed to zero at the inner boundary and to 1.0 at the outer boundary
//
// components 0 and 1 are the negative of the gradient of the potential satisfying the Laplace equation
// component 2 is the potential satisfying the Laplace equation
class ExactSolution : public Function<2>
{
public:
  ExactSolution() : Function<2>(3) {}
  double value (const Point<2>&p, const unsigned int component) const
  {
    if(component==0)
      return -1./log(2.) * p(0) / (p(0)*p(0)+p(1)*p(1));
    else if(component==1)
      return -1./log(2.) * p(1) / (p(0)*p(0)+p(1)*p(1));
    else
      return 1./log(2.) * log( sqrt(p(0)*p(0)+p(1)*p(1)) / 0.5 );
  }
};

//Mixed Laplace class (essentially copied and pasted from the deal.ii tutorials)
class MixedLaplace
{
public:
  MixedLaplace(const Mapping<2>& mapping);
  void run(const unsigned int refinements);
private:
  void make_grid(const unsigned int refinements);
  void setup_system();
  void assemble_system();
  void solve();
  void output_results(const unsigned int file_index) const;
  void difference_to_exact_solution() const;
  Triangulation<2>			triangulation;
  FESystem<2>					fe;
  DoFHandler<2>				dof_handler;
  SparsityPattern				sparsity_pattern;
  SparseMatrix<double>		system_matrix;
  Vector<double>				solution;
  Vector<double>				system_rhs;
  const SphericalManifold<2>	manifold;
  const Mapping<2>&			mapping;
};

MixedLaplace::MixedLaplace(const Mapping<2>& mapping)
  :
  //use quadratic elements for gradients of potential and linear elements for potential itself in order to satisfy inf-sup
  fe(FE_Q<2>(2), 2, FE_Q<2>(1), 1),
  dof_handler(triangulation),
  mapping(mapping)
{}

void MixedLaplace::make_grid(const unsigned int refinements)
{
  GridGenerator::hyper_shell(triangulation, Point<2>(), 0.5, 1, 8);
  for(const auto& cell : triangulation.active_cell_iterators())
    {
      for(unsigned int face=0; face<GeometryInfo<2>::faces_per_cell; ++face)
        {
          if(cell->face(face)->at_boundary()){
            const Point<2> point=cell->face(face)->center();
            double r=point.distance(Point<2>());
            if(r<0.75)
              cell->face(face)->set_boundary_id(0);
            else
              cell->face(face)->set_boundary_id(1);
          }
        }

    }
  triangulation.reset_all_manifolds();
  triangulation.set_all_manifold_ids(0);
  triangulation.set_all_manifold_ids_on_boundary(1);
  triangulation.set_manifold(1, manifold);
  triangulation.refine_global(refinements);
}

void MixedLaplace::setup_system()
{
  dof_handler.distribute_dofs(fe);
  DynamicSparsityPattern dsp(dof_handler.n_dofs());
  DoFTools::make_sparsity_pattern(dof_handler, dsp);
  sparsity_pattern.copy_from(dsp);
  system_matrix.reinit (sparsity_pattern);
  solution.reinit(dof_handler.n_dofs());
  system_rhs.reinit(dof_handler.n_dofs());
}
void MixedLaplace::assemble_system()
{
  QGauss<2> quadrature_formula(3);
  QGauss<1> face_quadrature_formula(3);
  FEValues<2> fe_values(mapping, fe, quadrature_formula,
                        update_values    | update_gradients |
                        update_quadrature_points  | update_JxW_values);
  FEFaceValues<2> fe_face_values(mapping, fe, face_quadrature_formula,
                                 update_values    | update_normal_vectors |
                                 update_quadrature_points  | update_JxW_values);
  const unsigned int dofs_per_cell = fe.dofs_per_cell;
  const unsigned int n_q_points = quadrature_formula.size();
  const unsigned int n_face_q_points = face_quadrature_formula.size();
  FullMatrix<double> cell_matrix(dofs_per_cell, dofs_per_cell);
  Vector<double>     cell_rhs(dofs_per_cell);
  vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
  const FEValuesExtractors::Vector velocities(0);
  const FEValuesExtractors::Scalar pressure(2);

  for(const auto &cell: dof_handler.active_cell_iterators())
    {
      cell_matrix = 0;
      cell_rhs = 0;
      fe_values.reinit(cell);
      //assemble volume terms
      for (unsigned int q=0; q<n_q_points; ++q)
        for (unsigned int i=0; i<dofs_per_cell; ++i)
          {
            const Tensor<1,2> phi_i_u = fe_values[velocities].value(i, q);
            const double  div_phi_i_u = fe_values[velocities].divergence(i, q);
            const double      phi_i_p = fe_values[pressure].value(i, q);
            for (unsigned int j=0; j<dofs_per_cell; ++j)
              {
                const Tensor<1,2> phi_j_u = fe_values[velocities].value(j, q);
                const double  div_phi_j_u = fe_values[velocities].divergence(j, q);
                const double  phi_j_p     = fe_values[pressure].value(j, q);
                cell_matrix(i,j) +=  (phi_i_u * phi_j_u
                                      - div_phi_i_u * phi_j_p
                                      - phi_i_p * div_phi_j_u)
                  * fe_values.JxW(q);
              }
          }
      //assemble boundary term to implement that potential is 1 at outer boundary
      for(unsigned int face_n=0; face_n<GeometryInfo<2>::faces_per_cell; ++face_n)
        if (cell->at_boundary(face_n) && cell->face(face_n)->boundary_id()==1)
          {
            fe_face_values.reinit (cell, face_n);
            for (unsigned int q=0; q<n_face_q_points; ++q)
              for (unsigned int i=0; i<dofs_per_cell; ++i)
                cell_rhs(i) += -(fe_face_values[velocities].value(i, q) *
                                 fe_face_values.normal_vector(q) *
                                 fe_face_values.JxW(q));
          }
      //distribute to global matrix and rhs
      cell->get_dof_indices(local_dof_indices);
      for(unsigned int i=0; i<dofs_per_cell; ++i)
        for (unsigned int j=0; j<dofs_per_cell; ++j)
          system_matrix.add(local_dof_indices[i],
                            local_dof_indices[j],
                            cell_matrix(i,j));
      for (unsigned int i=0; i<dofs_per_cell; ++i)
        system_rhs(local_dof_indices[i]) += cell_rhs(i);
    }
}

void MixedLaplace::solve()
{
  SparseDirectUMFPACK  A_direct;
  A_direct.initialize(system_matrix);
  A_direct.vmult(solution, system_rhs);
}

void MixedLaplace::output_results(const unsigned int file_index) const
{
  vector<string> solution_names(2, "u");
  solution_names.emplace_back("p");
  vector<DataComponentInterpretation::DataComponentInterpretation> interpretation(2, DataComponentInterpretation::component_is_part_of_vector);
  interpretation.push_back(DataComponentInterpretation::component_is_scalar);
  DataOut<2> data_out;
  data_out.add_data_vector (dof_handler, solution, solution_names, interpretation);
  data_out.build_patches(mapping, 2, DataOut<2>::curved_inner_cells);
  ofstream output("solution_" + 	Utilities::int_to_string(file_index) + ".vtk");
  data_out.write_vtk(output);
  cout << "  Solution written to=" << "solution_" + 	Utilities::int_to_string(file_index) + ".vtk" << endl;
}

void MixedLaplace::difference_to_exact_solution() const
{
  Vector<double> cell_differences(triangulation.n_active_cells());
  VectorTools::integrate_difference(mapping, dof_handler, solution, ExactSolution(), cell_differences, QGauss<2>(3), VectorTools::NormType::Linfty_norm);
  double e=VectorTools::compute_global_error 	(triangulation, cell_differences, VectorTools::NormType::Linfty_norm);
  cout << "  Error in L_infty norm=" << e << endl;
}

void MixedLaplace::run(const unsigned int refinements)
{
  make_grid(refinements);
  setup_system();
  assemble_system();
  solve();
  output_results(refinements);
  difference_to_exact_solution();
}

int main ()
{
  //if that's changed to MappingQGeneric, everything works fine
  const MappingQGeneric<2> mapping(2);

  //number of refinement steps (problem is solved for each refinement step in order to analyze the convergence properties)
  const unsigned int n_refinements=7;

  //do the computations
  for(unsigned int i1=0; i1<n_refinements+1; ++i1){
    MixedLaplace laplace_problem(mapping);
    cout << "Number of refinements: " << i1 << ":" << endl;
    laplace_problem.run(i1);
  }
  return 0;
}
```
If you still obtain some slight differences in results to mine, you should investigate the problem and see if there is any sensitivity on small changes (close to singular system?) - I certainly have some different libraries somewhere, but the results should be stable apart from roundoff.

**Comment by starki0815** [comment]
Thank you! You have assigned the flat manifold to the interior before refinement and I did it afterwards. This explains the difference and I can now reproduce your numbers.

As far as the rate of convergence is concerned, I have tried to interpolate the exact solution to the nodes instead of assembling and solving the system. The resulting rates of convergence of the interpolation error are quadratic for both cases (i.e., for the case with flat manifold in the interior as well as for the case with spherical manifold everywhere). So, my intuition has been wrong and the fact that there is a transition between a spherical manifold and a flat manifold within a single layer of element obviously doesn't affect interpolation estimates. Also, from the literature, there is the estimate `|| u-u_h || + || p-p_h || < C * ( inf || u-v_h || + inf || p-q_h || )` (e.g. [here](https://www.sciencedirect.com/science/article/pii/004578259090157H), Eq. (3.56)). In view of this result, it seems strange to me that I get only first order convergence for the case with flat manifold in the interior.

**Comment by bangerth** [comment]
It can't affect the interpolation error because interpolation happens locally on each cell independently of all other cells. So the interpolation operator knows nothing about a discontinuity in the gradient of the mapping between cells. But the Galerkin projection does, and so it is not impossible for this jump to show up in the constant `C` in your estimate. For you to get first order convergence, this would imply that `C ~ h^{-1}`, which is something you can get if `C` depends on the second derivatives of the mapping (globally); if the mapping is only once continuously differentiable, then its second derivative will grow with `h`. (Or maybe `C` depends on the first derivative of the mapping, but the mapping is really only continuous.)

**Comment by starki0815** [comment]
Thank you Wolfgang for your response and sorry for going off topic here!

Of course, you are right and interpolation can't be affected by the regularity of the mapping across cell boundaries. What I was rather thinking, is that the mapping used on an individual cell could affect the interpolation error on that cell; and that mapping does depend of course on the choice of the (mapping) support points (and, therefore, on the question whether `MappingQ` or `MappingQGeneric` is used).

Also, it is clear to me that if `C` would depend on `h`, this would blow up my argument regarding the rate of convergence. However, the reference cited above does say explicitly that `C` is independent of `h` and I can also not see a point in the proof where `h` could possibly enter `C`. But maybe I am getting something wrong there ...

**Comment by bangerth** [comment]
I'm not steeped enough in this area to recall the details, but the typical argument for an `h` to creep into a `C` is as follows: Let's say that the proof of convergence assumes that the gradient of the mapping stays bounded. This is true in all common cases, but you can break the assumption on triangles by choosing a particularly stupid refinement rule that divides triangles at the smallest angle of a cell, which consequently becomes smaller and smaller. A similar argument could be made for a refinement strategy that leads to triangles with one angle that tends to 180 degrees under refinement. 

Now, the "constant" `C` in the convergence proof really does depend on the smallest and largest angles of triangles in a mesh. Using reasonable refinement strategies, both of these angles remain bounded away from zero and 180 degrees, and consequently`C` really is a constant. But if you have a poor refinement strategy, then `C` would actually increase with mesh refinement. Depending on the exact way in which the minimal and maximal angle enter the constant `C`, and in which they deteriorate under mesh refinement, it is quite possible that you lose an order of convergence. For example, if `C=C_1 * 1/sin(alpha)` with `alpha` the minimal angle, and if `alpha` is proportional to `h`, then `C = O(h^{-1})`.

Similar arguments are conceivable (but further outside my realm of knowledge) with mappings with sharp transitions.

**Comment by starki0815** [comment]
Are you sure that the way how `h` could  enter `C` is something which has to do with the global regularity of the mapping? What my understandings of the matter was so far: If you distort the mesh cells in the way you describe it (if I recall it correctly, it's then called a mesh, which is not quasi-uniform), this will affect the (local) interpolation estimate and this is how `h` can enter the constant in the interpolation estimate and, consequently, the overall error estimate. Intuitively, I would expect that in the limit that `h`goes to `0`, the Jacobian matrix must tend to become a constant (i.e., it does not depend on position in a cell) on every cell of the mesh in order to allow for a transfer of the interpolation estimate made on the reference cell to the cell in real space. For the mesh I am using, this is clearly the case because cells tend to become rectangles in the limit that `h` goes to zero.

Now, as I have checked that the interpolation error shows the expected rate of convergence, it seems to me that `h` cannot enter `C` in the way you are describing it. Of course, if indeed the global regularity of the mapping matters, then you are right. But right now I don't understand how this could happen. Maybe there's an expert here, who could tell straight away?

In any case, if it turns out, that the sharp mapping indeed affects the rate of convergence in the way you are suspecting it, the documentation of MappingQGeneric saying 
> The interpolation scheme used in this class makes sure that curved descriptions can go over to flat descriptions within a single layer of elements, maintaining the overall optimal convergence rates of the finite element interpolation

should be modified (after all, the Stokes problem I am solving is nothing fancy ...)

**Comment by starki0815** [comment]
After some further experimenting I believe that the problem is neither directly related to the transition between manifolds nor is it some bug. Rather, it's probably that I am looking in the wrong space for solutions. Although Taylor-Hood works in practice for most problems with the mixed Laplacian, the pair Raviart-Thomas/DGQ is clearly the proper choice (in the way it is done in the tutorial step 20). I was aware of that, but I thought this wouldn't matter if the exact solution of my particular problem is within the space I am using. Apparently that's wrong. If I use the Raviart-Thomas/DGQ, convergence is exactly as expected for all cases. So sorry again for the long off topic discussion.
